python/tank/descriptor/io_descriptor/downloadable.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import shutil
+
+from .base import IODescriptorBase
+from ..errors import TankDescriptorError, TankError
+from ...util import filesystem
+
+from ... import LogManager
+log = LogManager.get_logger(__name__)
+
+
+class IODescriptorDownloadable(IODescriptorBase):
+    """
+    Base class for descriptors that can be downloaded locally to a path on disk.
+    """
+    def download_local(self):
+
+        # Return if the descriptor exists locally.
+        if self.exists_local():
+            return
+
+        # cache into a temporary location
```

Comment:
we should add debug logging here, so that a (support) person looking at a debug log is able to follow along and see what is going on i/o-wise. A good way to test this is to put your TD hat on, turn on debug and see what kind of output you would think would be helpful to see if you were remote diagnosing some weird issue around descriptor downloads! Most likely i/o related info, but perhaps more things.